mlflow/models/evaluation/base.py

```diff
@@ -74,8 +74,8 @@ class EvaluationMetric:
 
             def eval_fn(
                 eval_df: Union[pandas.Dataframe, pyspark.sql.DataFrame],
-                builtin_metrics: Dict[str, float],
-            ) -> float:
+                builtin_metrics: Dict[str, MetricValue],
+            ) -> MetricValue:
```

Comment:
To keep it backwards compatible with Traditional ML. Should this be a union of float and MetricValue?

Reply:
The builtin_metrics type or the return type? We are passing in the builtin_metrics so it will have to be MetricValue - I was thinking that we could accept float return value [here](https://github.com/mlflow/mlflow/pull/9440/files#r1307938094) in the validation but we don't need to write it in the docs. What do you think?

Reply:
It could be something like this:

```
 def eval_fn(
                eval_df: Union[pandas.Dataframe, pyspark.sql.DataFrame],
                builtin_metrics: Dict[str, Union[float, MetricValue]],
            ) -> Union[float, MetricValue]:
```

So it would work for both use cases? Thoughts?

Reply:
We can definitely have Union[float, MetricValue] as the return type. 
Since we will be calculating/passing in the builtin_metrics in mlflow.evaluate(), are you proposing that we pass in the already existing metrics as floats and future/new metrics as MetricValue? I feel like this could be confusing for users because they will have to know/figure out the type of each builtin metric inside their eval_fn. On the other hand, if we make them all MetricValue it will break existing eval_fn for custom metrics.

Reply:
@prithvikannan @dbczumar Opinions here?

Reply:
Pretty tricky for users to consume. They need do `if isinstance(...)` in every `eval_fn` to avoid errors. I'm in favor of always using an object that encapsulates the aggregates and the scores of the built-ins. We need to be backwards compatible with existing eval functions, though, so I suggest we deprecate this field and insert a new one with a different name, e.g. just `metrics`. Thoughts @annzhang-db @prithvikannan @sunishsheth2009  ?

Reply:
Agree with @dbczumar -- mixing floats with MetricValues will be confusing to users. Deprecating seems like the right path forward to me...what additional complexity comes with this?

Reply:
Not much additional complexity.

We can start asking users to pass an `eval_fn` that takes `eval_df, metrics: Dict[str, MetricValue]` instead of `eval_df, builtin_metrics: DIct[str, float]` and use `inspect.signature()` to read the type hints If the variable type hint isn't `Dict[str, MetricValue]`, pass `builtin_metrics` and print a warning. Else, pass the updated `metrics` of type `Dict[str, MetricValue]`. WDYT?

Reply:
This makes sense to me, in that case should we also write/document that eval_fn should return MetricValue (instead of Union[float, MetricValue]) but still allow functions that returns float to work as they have in the past (with a warning as well?)

Reply:
Updated to have `metrics: Dict[str, MetricValue]` and return `MetricValue` from eval_fn but all old eval_fn should still work.

Reply:
`eval_fn` should be able to *return* `Union[float, MetricValue]`, but we should recommend that it *accept* `Dict[str, MetricValue]`.